### ISSUE TITLE
dasdcopy64 compressed CFBA64 --> uncompressed FBA64 not working

### ISSUE THREAD

**Issue by bluecmd** [SDL-Hercules-390/hyperion#212]
Hi,

This is at c66079f431ba0f2bb7adcdf4d3c4bf13ce7cd19a (current HEAD as of this writing).

I have a compressed FBA64 that I am trying to create a flat image of.

```
$ dasdcopy64  -lfs -o FBA64 vmcom.fba vmcom.img
HHC02499I Hercules utility dasdcopy64 - 64-bit DASD copy/convert - version 4.2.0.0-SDL-gc66079f4
HHC01414I (C) Copyright 1999-2019 by Roger Bowler, Jan Jaeger, and others
HHC01417I ** The SoftDevLabs version of Hercules **
HHC01415I Build date: May 25 2019 at 16:44:16
HHC00507I 0:0000 FBA file vmcom.fba: origin 0, blks 20971520
HHC00473I 0:0000 CKD64 file vmcom.img: creating 3370 volume : 20971520 sectors, 512 bytes/sector
HHC00471I 0:0000 CKD64 file vmcom.img: 20971520 sectors successfully written
HHC00507I 0:0001 FBA file vmcom.img: origin 0, blks 4194304
/  20%   34953HHC00502E 0:0001 FBA file vmcom.img: error in function read(): unexpected end of file
HHC00007I Previous message from function 'fba_dasd_read_blkgrp' at fbadasd.c(654)
HHC02434E Write error on file vmcom.img: block 34953 stat=0E
HHC00007I Previous message from function 'main' at dasdcopy64.c(507)
```

The file is 10G so 20% would be suspiciously close to 2G, i.e. 31 bit signed wrap around.

To reproduce:

1. dasdinit64 -0 vmcom.fba 9336 'VMCOM1' $((2*1024*1024*10))
2. Format the drive and write stuff on it. I format it using z/VM installation program.
3. dasdcopy64  -lfs -o FBA64 vmcom.fba vmcom.img

**Comment by Fish-Git** [comment]
Signed integer overflow. Two places:

First one in `dasdcopy64` itself where it's calculating an unsigned 64-bit "number of bytes remaining" value by multiplying two 32-bit signed integer values without casting one of them to a 64-bit signed integer beforehand. Second one is in `dasdutil64` where it's making the exact same mistake at the point where it's calculating the file size to be used in an `ftruncate` call to create the output file.

I'll be committing the fix shortly.  Thanks for reporting it.  It's something I completely missed during my CCKD64 effort.

(I wish there was some compiler option I could enable that would report such potential bugs.)

**Comment by Fish-Git** [comment]
There was actually yet another bug that was involved too that was the root cause of the `unexpected end of file` error: the FBA device handler only supports the reading/writing multiples of 120 sectors (known as a "block group"), and 20971520 sectors is not a multiple of 120, so the output file that was created was slightly too small, leading to the `unexpected end of file` error. (The signed integer overflow was causing the output file to be created _way_ too small, but even with that fixed, the error was still occurring due to the non-integral block group size requested number of sectors.)

All three bug have now been fixed. But it's late and I'm tired. I'll commit the fix tomorrow. Sorry for the delay.

**Comment by bluecmd** [comment]
Thanks Fish! Great work as always!

**Comment by Fish-Git** [comment]
Fixed by commit e9968603f6b8d3083cbc34ca20aefeff2f5145f0.

Closing...
